### PR TITLE
Plugin Extensions: Require meta-data to be defined in `plugin.json` during development mode

### PR DESCRIPTION
**Pull request by leventebalogh**
**Related:** https://github.com/grafana/grafana/issues/92005

**This PR doesn't change anything for plugins running in production.**
**This PR doesn't affect core Grafana extension points.**

### Goals
We would like to decrease the discrepancy between the meta-info recorded in the `plugin.json` for extensions and what we have in the code. The aim of this PR is to only allow extensions **in dev mode** that are registered correctly in the plugin's `plugin.json`.

Screenshot 2024-09-20 at 9 10 37

**Why is this important?**
We would like to be able to only load app plugins when it's necessary, compared to loading all of them in advance (how it is now). In order to achieve that in a reliable way we need to understand what the plugins are extending and what extension points they provide.

### What changed?
- [x] _Backend:_ Update the go models for plugins - add extensions meta-data from the plugin.json
- [x] _Backend:_ Expose extensions related plugin meta-data to the frontend
- [x] _Backend:_ Add a fallback for old plugin.json schemas
- [x] _Frontend:_ Add extensions meta-data validation when in dev-mode
- [x] _Frontend:_ Add and update tests

### CODE REVIEW

File: packages/grafana-data/src/types/plugin.ts [hunk 1]
```diff
@@ -111,6 +112,40 @@ export interface PluginDependencies {
...
+  };
+}
+
+export interface PluginExtensions {
+  // The component extensions that the plugin registers
+  addedComponents: Array<{
+    targets: string[];
+    title: string;
+    description: string;
+  }>;
```
Comment (mckn) [inline packages/grafana-data/src/types/plugin.ts:127]: nit: would it make sense to introduce a type here? 

```ts
type ExtensionInfo = {
   targets: string[];
   title: string;
   description;
}
```

File: pkg/services/store/testdata/public_testdata.golden.jsonc
Comment (mckn) [inline pkg/services/store/testdata/public_testdata.golden.jsonc:1]: Is this change intentional?

**Review by mckn — APPROVED** [review]
LGTM! Great work with this Levi ⭐

File: pkg/plugins/models.go [hunk 2]
```diff
@@ -42,9 +43,100 @@ func (e DuplicateError) Is(err error) bool {
...
+					Targets:     []string{extensionV1.ExtensionPointID},
+					Title:       extensionV1.Title,
+					Description: extensionV1.Description,
+				}
+
+				e.AddedComponents = append(e.AddedComponents, extensionV2);
+			}
+		}
+
+		return nil;
```
Comment (sunker) [inline pkg/plugins/models.go:108]: Correct med if I'm wrong, but this fallback will not take us all the way since the legacy schema doesn't support extension points? So if we silently migrate this, things may still not work as expected in the app. Wouldn't it be better to simply drop support for extensions V1 format and always require the new schema format?

File: public/app/features/plugins/extensions/registry/ExportedComponentsRegistry.test.ts [hunk 3]
```diff
@@ -1,15 +1,62 @@
...
+        exposedComponents: [],
+      },
+    },
+    extensions: {
+      addedLinks: [],
+      addedComponents: [],
+      exposedComponents: [],
+      extensionPoints: [],
+    },
+  };
```
Comment (sunker) [inline public/app/features/plugins/extensions/registry/ExportedComponentsRegistry.test.ts:47]: Since this mock is being used in a few places, maybe worth adding it to a `__mocks__` dir?

File: public/app/features/plugins/extensions/utils.tsx [hunk 4]
```diff
@@ -408,3 +411,145 @@ export const openAppInSideview = (pluginId: string) => sidecarService.openApp(pl
...
+// Comes from the `app_mode` setting in the Grafana config (defaults to "development")
+// Can be set with the `GF_DEFAULT_APP_MODE` environment variable
+export const isGrafanaDevMode = () => config.buildInfo.env === 'development';
+
+// Checks if the meta information is missing from the plugin's plugin.json file
+export const isExtensionPointMetaInfoMissing = (extensionPointId: string, pluginContext: PluginContextType) => {
+  const pluginId = pluginContext.meta?.id;
+  const extensionPoints = pluginContext.meta?.extensions?.extensionPoints;
+
+  if (!extensionPoints || !extensionPoints.find((ep) => ep.id === extensionPointId)) {
```
Comment (sunker) [inline public/app/features/plugins/extensions/utils.tsx:424]: nit
```suggestion
  if (!extensionPoints || !extensionPoints.some((ep) => ep.id === extensionPointId)) {
```

File: public/app/features/plugins/extensions/validators.ts [hunk 5]
```diff
@@ -53,12 +53,31 @@ export function isLinkPathValid(pluginId: string, path: string) {
...
+  // If a `pluginId` is provided, the extension point should be prefixed with the plugin id.
+  if (pluginId) {
+    return Boolean(extensionPointId.startsWith(`plugins/${pluginId}/`) || extensionPointId.startsWith(`${pluginId}/`));
+  }
+
+  if (extensionPointId.startsWith('grafana/')) {
+    return true;
+  }
+
+  return Boolean(extensionPointId.replace(/^plugins\//, '').match(regex));
```
Comment (sunker) [inline public/app/features/plugins/extensions/validators.ts:80]: 👏

**Review by sunker — COMMENTED** [review]
This is a great step forward! Nice work on this. Added a few minor comments. Also haven't tested this yet, but will do during the afternoon.

Q: Would it be helpful to add a warning in case someone registers a link/component to an extension point that is not defined by any of the apps that are currently installed on the instance?

File: pkg/services/store/testdata/public_testdata.golden.jsonc
Comment (leventebalogh) [reply pkg/services/store/testdata/public_testdata.golden.jsonc:1]: Nice catch, I think no, let me have a look at it 🙏

File: packages/grafana-data/src/types/plugin.ts [hunk 6]
```diff
@@ -111,6 +112,40 @@ export interface PluginDependencies {
...
+
+  // The React components that the plugin exposes
+  exposedComponents: Array<{
+    id: string;
+    title: string;
+    description: string;
```
[the remark this answers is not in the document]
Comment (leventebalogh) [reply packages/grafana-data/src/types/plugin.ts:140]: Nice catch 👍

File: pkg/plugins/models.go (hunk 2 above)
Comment (leventebalogh) [reply pkg/plugins/models.go:108]: No no, you are right that it's not gonna take us all the way.

This backwards compatibility is only in place, so any production deploys of the internal app plugins won't break once they receive this update (we have previously updated their plugin.json to contain the now outdated `extensions: []` meta info. We could see and specify what Grafana versions are supported in cloud, so this is more just to be safe. 

I hope that the following two things will help them to update:
- the plugin.json schema only supports the new version and will throw warnings
- if they introduce an extension point (and have the old plugin.json schema), then in dev mode it won't work for them

Did that answer your question mate?

File: public/app/features/plugins/extensions/registry/ExportedComponentsRegistry.test.ts (hunk 3 above)
Comment (leventebalogh) [reply public/app/features/plugins/extensions/registry/ExportedComponentsRegistry.test.ts:47]: I think that's a good point 👍

File: pkg/services/store/testdata/public_testdata.golden.jsonc
Comment (leventebalogh) [reply pkg/services/store/testdata/public_testdata.golden.jsonc:1]: Removed in https://github.com/grafana/grafana/pull/93429/commits/c041657fe361f3fdaaa12319334065a68cb3b550.

**Comment by leventebalogh** [comment]
@sunker 
> Q: Would it be helpful to add a warning in case someone registers a link/component to an extension point that is not defined by any of the apps that are currently installed on the instance?

File: pkg/plugins/models.go (hunk 2 above)
Comment (sunker) [reply pkg/plugins/models.go]: I see, thanks for clarifying!

**Review by sunker — COMMENTED** [review]
Have done some testing now. Could only find one issue, and that is that the now deprecated `usePluginExtensions` hook still works even if the extension point is missing in plugin.json. 

Pushed some code that adds a new nested app that tries to register extensions and an extension point but without declaring them in package json. Added a few tests that verifies that they're **not** functioning properly in dev mode due to missing declarations. 

Think we could more to harmonize error messages and warnings so the become easier to read, but probably better to address that in a follow up PR where console.log is replaced with using the new extensions debug log.

File: packages/grafana-data/src/types/plugin.ts (hunk 1 above)
Comment (leventebalogh) [reply packages/grafana-data/src/types/plugin.ts]: Should be fixed in https://github.com/grafana/grafana/pull/93429/commits/a721c7ca42ea28f36df7b525bdea64bcc1adf9ce 👍

File: packages/grafana-data/src/types/plugin.ts [hunk 7]
```diff
@@ -111,6 +112,40 @@ export interface PluginDependencies {
...
+  // The component extensions that the plugin registers
+  addedComponents: Array<{
+    targets: string[];
+    title: string;
+    description: string;
+  }>;
+
+  // The link extensions that the plugin registers
+  addedLinks: Array<{
+    targets: string[];
```
[the remark this answers is not in the document]
Comment (leventebalogh) [reply packages/grafana-data/src/types/plugin.ts]: Should be fixed in https://github.com/grafana/grafana/pull/93429/commits/a721c7ca42ea28f36df7b525bdea64bcc1adf9ce.

File: packages/grafana-data/src/types/plugin.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (leventebalogh) [reply packages/grafana-data/src/types/plugin.ts]: Should be fixed in https://github.com/grafana/grafana/pull/93429/commits/a721c7ca42ea28f36df7b525bdea64bcc1adf9ce. 👍

File: public/app/features/plugins/extensions/utils.tsx (hunk 4 above)
Comment (leventebalogh) [reply public/app/features/plugins/extensions/utils.tsx]: Fixed in https://github.com/grafana/grafana/pull/93429/commits/d02fbf48df3716e7fdd4d6468de66d23b9ba8976.

File: public/app/features/plugins/extensions/usePluginLinks.tsx [hunk 8]
```diff
@@ -24,9 +28,32 @@ export function usePluginLinks({
...
+        `Extension point usePluginLinks("${extensionPointId}") - the id should be prefixed with your plugin id ("${pluginId}/").`
+      );
+      return {
+        isLoading: false,
+        links: [],
+      };
+    }
+
+    if (enableRestrictions && isExtensionPointMetaInfoMissing(extensionPointId, pluginContext)) {
+      logWarning(`usePluginLinks("${extensionPointId}") - The extension point is missing from the "plugin.json" file.`);
```
[the remark this answers is not in the document]
Comment (leventebalogh) [reply public/app/features/plugins/extensions/usePluginLinks.tsx]: Sure, addressed in https://github.com/grafana/grafana/pull/93429/commits/fc3a9fa4260204a212f433d1a4faee8f03795127 👍

File: public/app/features/plugins/extensions/registry/ExportedComponentsRegistry.test.ts (hunk 3 above)
Comment (leventebalogh) [reply public/app/features/plugins/extensions/registry/ExportedComponentsRegistry.test.ts:47]: Hmm, actually, I am being a little bit puzzled here. How are we using the `__mocks__` dir usually? If we use it for manual Jest mocks (mocking a module) then I guess this wouldn't really fit there? (It's not a module, but part of the config data structure). Or would you imagine just putting it in a JSON file and requiring it in from the different test cases that depend on it?

Comment (sunker) [reply public/app/features/plugins/extensions/registry/ExportedComponentsRegistry.test.ts:47]: Ah that's true. Moving it to its own file sounds like a good idea. 👍

**Review by torkelo — COMMENTED** [review]
Nice!

**Review by sunker — APPROVED** [review]
Tested with dev and prod builds, and appears to work as expected. Fantastic work on this @leventebalogh!

**Comment by leventebalogh** [comment]
Thanks for the testing @sunker 🙏
